core/src/ldml/ldml_transforms.cpp

```diff
@@ -6,64 +6,191 @@
 */
 
 #include "ldml_transforms.hpp"
+#include "debuglog.h"
+
+#ifndef assert
+#define assert(x) // TODO-LDML
+#endif
 
 namespace km {
 namespace kbp {
 namespace ldml {
 
-transforms::transforms() : simple() {
+
+transform_entry::transform_entry(const std::u16string &from, const std::u16string &to) : fFrom(from), fTo(to) {
+}
+
+size_t
+transform_entry::match(const std::u16string &input) {
+  if (input.length() < fFrom.length()) {
+    return 0;
+  }
+  // string at end
+  auto substr = input.substr(input.length() - fFrom.length(), fFrom.length());
+  if (substr != fFrom) {
+    return 0;
+  }
+  return substr.length();
+}
+
+std::u16string
+transform_entry::apply(const std::u16string &/*input*/, size_t /*matchLen*/) {
+  return fTo;
+}
+
+transforms::transforms() : transform_groups() {
 }
 
 void
-transforms::add(const string_list &list, const std::u16string &tostr) {
-  simple.insert({list, tostr});
+transforms::addTransformGroup(const transform_group &s) {
+  transform_groups.push_back(s);
 }
 
 size_t
-transforms::matchContext(const string_list &ctxt, std::u16string &outputString) const {
-  if (simple.size() == 0) {
-    return 0;  // no transforms, no match
-  }
+transforms::apply(const std::u16string &input, std::u16string &output) {
+  /**
+   * Accumulate the maximum matched size of the end of 'input' here.
+   * It will be replaced with 'output'.
+   * Matched can increment.
+   * TODO-LDML:  As an optimization, we could decrement 'matched' for every char of output
+   * which is already in input. Example (regex example):
+   * - str = "xxyyzz";
+   * - s/zz$/z/ -> match=2, output='z'
+   * - s/z$/zz/ -> match=2, output='zz'
+   *      (but could contract to match=0, output='' as already present)
+   * - s/zz$/zw/ -> match=2, output='zw'
+   *      (but could contract to match=1, output='w')
+   */
+  size_t matched              = 0;
+  std::u16string updatedInput = input;  // modified input
+  for (auto group = transform_groups.begin(); group < transform_groups.end(); group++) {
+    // for each transform group
+    // break out once there's a match
```

Comment:
We don't seem to be breaking out of the inner loop with first match. A group should only be applying the first matching transform. A cleaner structure would be to search the group for a matching transform, and then do the application of that found transform outside the inner loop.

pseudocode:

```
transform = group.findFirstMatch(updatedInput, &subMatched);
if(transform) {
  ... do the transform
}
```

or 

```
transform = findFirstMatchInGroup(group, updatedInput, &subMatched);
if(transform) {
  ... do the transform
}
```

Reply:
Have not restructured. but I did make it break out of the inner loop.